partition/ubootenv/env.go

```diff
@@ -0,0 +1,291 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ubootenv
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// FIXME: add config option for that so that the user can select if
+//        he/she wants env with or without flags
+var headerSize = 5
+
+// Env contains the data of the uboot environment
+type Env struct {
+	fname string
+	size  int
+	data  map[string]string
+}
+
+// little endian helpers
+func readUint32(data []byte) uint32 {
+	var ret uint32
+	buf := bytes.NewBuffer(data)
+	binary.Read(buf, binary.LittleEndian, &ret)
+	return ret
+}
+
+func writeUint32(u uint32) []byte {
+	buf := bytes.NewBuffer(nil)
+	binary.Write(buf, binary.LittleEndian, &u)
+	return buf.Bytes()
+}
+
+// Create a new empty uboot env file with the given size
+func Create(fname string, size int) (*Env, error) {
+	f, err := os.Create(fname)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	env := &Env{
+		fname: fname,
+		size:  size,
+		data:  make(map[string]string),
+	}
+
+	return env, nil
+}
+
+// OpenFlags instructs open how to alter its behavior.
+type OpenFlags int
+
+const (
+	// OpenBestEffort instructs OpenWithFlags to skip malformed data without returning an error.
+	OpenBestEffort OpenFlags = 1 << iota
+)
+
+// Open opens a existing uboot env file
+func Open(fname string) (*Env, error) {
+	return OpenWithFlags(fname, OpenFlags(0))
+}
+
+// OpenWithFlags opens a existing uboot env file, passing additional flags.
+func OpenWithFlags(fname string, flags OpenFlags) (*Env, error) {
+	f, err := os.Open(fname)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	contentWithHeader, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	crc := readUint32(contentWithHeader)
+
+	payload := contentWithHeader[headerSize:]
+	actualCRC := crc32.ChecksumIEEE(payload)
+	if crc != actualCRC {
+		return nil, fmt.Errorf("bad CRC: %v != %v", crc, actualCRC)
+	}
+	eof := bytes.Index(payload, []byte{0, 0})
+
+	data, err := parseData(payload[:eof], flags)
+	if err != nil {
+		return nil, err
+	}
+
+	env := &Env{
+		fname: fname,
+		size:  len(contentWithHeader),
+		data:  data,
+	}
+
+	return env, nil
+}
+
+func parseData(data []byte, flags OpenFlags) (map[string]string, error) {
+	out := make(map[string]string)
+
+	for _, envStr := range bytes.Split(data, []byte{0}) {
+		if len(envStr) == 0 || envStr[0] == 0 || envStr[0] == 255 {
+			continue
+		}
+		l := strings.SplitN(string(envStr), "=", 2)
+		if len(l) != 2 {
+			if flags&OpenBestEffort == OpenBestEffort {
+				continue
+			}
+			return nil, fmt.Errorf("cannot parse line %q as key=value pair", envStr)
+		}
+		key := l[0]
```

Comment:
shouldn't we check that key is not empty?  especially given  we panic in iterEnv in such a case?